src/target/stm32g0.c

```diff
@@ -0,0 +1,758 @@
+	{ "erase_mass", (cmd_handler)stm32g0_cmd_erase_mass,
+	  "Erase entire flash memory" },
+	{ "erase_bank1", (cmd_handler)stm32g0_cmd_erase_bank1,
+	  "Erase entire bank1 flash memory" },
+	{ "erase_bank2", (cmd_handler)stm32g0_cmd_erase_bank2,
+	  "Erase entire bank2 flash memory" },
+	{ "option", (cmd_handler)stm32g0_cmd_option,
+	  "Manipulate option bytes" },
+	{ "irreversible", (cmd_handler)stm32g0_cmd_irreversible,
+	  "Allow irreversible operations: (enable|disable)" },
+	{ NULL, NULL, NULL }
+};
+
+static void stm32g0_add_flash(target *t, uint32_t addr, size_t length,
+			      size_t blocksize)
+{
+	struct target_flash *f = calloc(1, sizeof(*f));
+	if (!f) { /* calloc failed: heap exhaustion */
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	f->start = addr;
+	f->length = length;
+	f->blocksize = blocksize;
+	f->erase = stm32g0_flash_erase;
+	f->write = stm32g0_flash_write;
+	f->buf_size = FLASH_PAGE_SIZE;
+	f->erased = 0xFF;
+	target_add_flash(t, f);
+}
+
+/*
+ * Probe for a known STM32G0 MCU.
+ */
+bool stm32g0_probe(target *t)
+{
+	switch (t->idcode) {
+	case STM32G03_4:
+		/* SRAM 8 kB, Flash up to 64 kB */
+		t->driver = "STM32G03/4";
+		break;
+	case STM32G05_6:
+		/* SRAM 18 kB, Flash up to 64 kB */
+		t->driver = "STM32G05/6";
+		break;
+	case STM32G07_8:
+		/* SRAM 36 kB, Flash up to 128 kB */
+		t->driver = "STM32G07/8";
+		break;
+	case STM32G0B_C:
+		/* SRAM 144 kB, Flash up to 512 kB */
+		t->driver = "STM32G0B/C";
+		break;
+	default:
+		return false;
+	}
+
+	t->attach = stm32g0_attach;
+	t->detach = stm32g0_detach;
+	target_add_commands(t, stm32g0_cmd_list, t->driver);
+	return true;
+}
+
+/*
+ * In addition to attaching the debug core with cortexm_attach(), this function
+ * keeps the FCLK and HCLK clocks running in Standby and Stop modes while
+ * debugging.
+ * The watchdogs (IWDG and WWDG) are stopped when the core is halted. This
+ * allows basic Flash operations (erase/write) if the watchdog is started by
+ * hardware or by a previous program without prior power off. Particularly
+ * useful with hosted blackmagic binary write where the debug registers cannot
+ * be written.
+ * Populate the memory map and add custom commands.
+ */
+static bool stm32g0_attach(target *t)
+{
+	uint32_t ram_size = 0U;
+
+	if (!cortexm_attach(t))
+		return false;
+
+	saved_regs.rcc_apbenr1 = target_mem_read32(t, RCC_APBENR1);
+	target_mem_write32(t, RCC_APBENR1, saved_regs.rcc_apbenr1 |
+	                   RCC_APBENR1_DBGEN);
+	saved_regs.dbg_cr = target_mem_read32(t, DBG_CR);
+	target_mem_write32(t, DBG_CR, saved_regs.dbg_cr |
+			   (DBG_CR_DBG_STANDBY | DBG_CR_DBG_STOP));
+	saved_regs.dbg_apb_fz1 = target_mem_read32(t, DBG_APB_FZ1);
+	target_mem_write32(t, DBG_APB_FZ1, saved_regs.dbg_apb_fz1 |
+	                   (DBG_APB_FZ1_DBG_IWDG_STOP | DBG_APB_FZ1_DBG_WWDG_STOP));
+
+	target_mem_map_free(t);
+
+	size_t flash_size = (size_t)target_mem_read16(t, FLASH_MEMORY_SIZE);
```

Comment:
I think it is best to announce the biggest flash size for that chip. Smaller flash sizes only means that that flash up to the real size is not tested or perhaps disabled. By using the largest size, we give the user control what to do with flash above the announced size. Do you have G0 devices with criplled flash? Can you test?

Reply:
I got an STM32G031 (32 kiB) and I can confirm that I succeeded in writing above 0x8000!
I implemented the full 64 kiB announcement for single bank devices.